File: src/main/java/com/gary/GalacticTrading/parser/QueryParser.java
package com.gary.GalacticTrading.parser;

import com.gary.GalacticTrading.utils.InputRegEx;
import com.gary.GalacticTrading.utils.QueryConstants;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;

import java.util.Arrays;
import java.util.Set;

/**
 * QueryParser class is responsible for parsing queries.
 */
@Component
@Slf4j
public class QueryParser {
    public String[] parseQuery(final String query) {
        if (query == null || query.isEmpty() || !query.matches(InputRegEx.QUERY)) {
            log.error(QueryConstants.INVALID_QUERY);
            return new String[] {QueryConstants.INVALID_QUERY};
        }
        String[] metalValueQuery = Arrays.stream(query.split(" ")).filter(s -> !s.equals("")).toArray(String[]::new);
        final int indexOfIs = Arrays.asList(metalValueQuery).indexOf("is");
        final int indexOfQuestionMark = Arrays.asList(metalValueQuery).indexOf("?");
        String[] unitMetalQuery = Arrays.copyOfRange(metalValueQuery, indexOfIs + 1, indexOfQuestionMark);

        log.debug("Parsed query: query -> {}", String.join(" ", unitMetalQuery));
        return unitMetalQuery;
    }
}

File: src/main/kotlin/com/gary/GalacticTrading/parser/QueryParser.kt
package com.gary.GalacticTrading.parser

import com.gary.GalacticTrading.utils.InputRegEx
import com.gary.GalacticTrading.utils.QueryConstants
import mu.KotlinLogging
import org.springframework.stereotype.Component
import java.util.*

/**
 * QueryParser class is responsible for parsing queries.
 */
@Component
class QueryParser {
    private val log = KotlinLogging.logger {}

    fun parseQuery(query: String?): Array<String> {
        if (query == null || query.isEmpty() || !query.matches(InputRegEx.QUERY.toRegex())) {
            log.error(QueryConstants.INVALID_QUERY)
            return arrayOf(QueryConstants.INVALID_QUERY)
        }
        val metalValueQuery: Array<String> = query.split(" ".toRegex()).filter { it != "" }.toTypedArray()
        val indexOfIs = metalValueQuery.indexOf("is")
        val indexOfQuestionMark = metalValueQuery.indexOf("?")
        val unitMetalQuery = Arrays.copyOfRange(metalValueQuery, indexOfIs + 1, indexOfQuestionMark)

        log.debug("Parsed query: query -> {}", java.lang.String.join(" ", *unitMetalQuery))
        return unitMetalQuery
    }
}
